requests.go

```diff
@@ -1,91 +1,61 @@
+func (c *Client) req(method, path string, body io.Reader, intercept func(*http.Request)) (rs *http.Response, err error) {
+	var redo bool
 	var r *http.Request
-	var retryBuf io.Reader
-
-	if body != nil {
-		// If the authorization fails, we will need to restart reading
-		// from the passed body stream.
-		// When body is seekable, use seek to reset the streams
-		// cursor to the start.
-		// Otherwise, copy the stream into a buffer while uploading
-		// and use the buffers content on retry.
-		if sk, ok := body.(io.Seeker); ok {
-			if _, err = sk.Seek(0, io.SeekStart); err != nil {
-				return
-			}
-			retryBuf = body
-		} else {
-			buff := &bytes.Buffer{}
-			retryBuf = buff
-			body = io.TeeReader(body, buff)
-		}
-		r, err = http.NewRequest(method, PathEscape(Join(c.root, path)), body)
-	} else {
-		r, err = http.NewRequest(method, PathEscape(Join(c.root, path)), nil)
-	}
+	var uri = PathEscape(Join(c.root, path))
+	auth, body := c.auth.NewAuthenticator(body)
+	defer auth.Close()
 
-	if err != nil {
-		return nil, err
-	}
-
-	for k, vals := range c.headers {
-		for _, v := range vals {
-			r.Header.Add(k, v)
+	for { // TODO auth.continue() strategy(true|n times|until)?
+		if r, err = http.NewRequest(method, uri, body); err != nil {
+			return
 		}
-	}
-
-	// make sure we read 'c.auth' only once since it will be substituted below
-	// and that is unsafe to do when multiple goroutines are running at the same time.
-	c.authMutex.Lock()
-	auth := c.auth
-	c.authMutex.Unlock()
 
-	auth.Authorize(r, method, path)
+		for k, vals := range c.headers {
+			for _, v := range vals {
+				r.Header.Add(k, v)
+			}
+		}
 
-	if intercept != nil {
-		intercept(r)
-	}
+		if err = auth.Authorize(c.c, r, path); err != nil {
+			return
+		}
 
-	if c.interceptor != nil {
-		c.interceptor(method, r)
-	}
+		if intercept != nil {
+			intercept(r)
+		}
 
-	rs, err := c.c.Do(r)
-	if err != nil {
-		return nil, err
-	}
+		if c.interceptor != nil {
+			c.interceptor(method, r)
+		}
 
-	if rs.StatusCode == 401 && auth.Type() == "NoAuth" {
-		wwwAuthenticateHeader := strings.ToLower(rs.Header.Get("Www-Authenticate"))
-
-		if strings.Index(wwwAuthenticateHeader, "digest") > -1 {
-			c.authMutex.Lock()
-			c.auth = &DigestAuth{auth.User(), auth.Pass(), digestParts(rs)}
-			c.authMutex.Unlock()
-		} else if strings.Index(wwwAuthenticateHeader, "basic") > -1 {
-			c.authMutex.Lock()
-			c.auth = &BasicAuth{auth.User(), auth.Pass()}
-			c.authMutex.Unlock()
-		} else {
-			return rs, newPathError("Authorize", c.root, rs.StatusCode)
+		if rs, err = c.c.Do(r); err != nil {
+			return
 		}
 
-		// retryBuf will be nil if body was nil initially so no check
-		// for body == nil is required here.
-		return c.req(method, path, retryBuf, intercept)
-	} else if rs.StatusCode == 401 {
-		return rs, newPathError("Authorize", c.root, rs.StatusCode)
+		if redo, err = auth.Verify(c.c, rs, path); err != nil {
+			io.Copy(io.Discard, rs.Body)
```

Comment:
Is it really necessary to copy the body to `io.Discard`? Just would like to know why this is done here, just for interest. ^^

Reply:
Not sure, if that is actually needed because the `Body.Close()` method does some things to try and read/discard the body (up to a limit size) in order for the TCP connection to be re-used.
https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/net/http/transfer.go;l=1002